src/robusta/core/sinks/common/channel_transformer.py

```diff
@@ -0,0 +1,94 @@
+ANNOTATIONS_PREF_PATTERN = regex.compile("\$?" + ESCAPED_ANNOTATIONS_PREF + "[\w.]+")  # noqa: W605
+BRACKETS_PATTERN = regex.compile(r"\$\{[^\}]+\}")
+COMPOSITE_PATTERN = r".*\$({?labels.[^$]+|{?annotations.[^$]+|{?cluster_name).*"
+ONLY_VALUE_PATTERN = r"^(labels.[^$]+|annotations.[^$]+|cluster_name)$"
+MISSING = "<missing>"
+
+
+class ChannelTransformer:
+    @classmethod
+    def validate_channel_override(cls, v: Union[str, None]):
+        if v:
+            if regex.match(ONLY_VALUE_PATTERN, v):
+                return "$" + v
+            if not regex.match(COMPOSITE_PATTERN, v):
+                err_msg = (
+                    f"channel_override must be '{CLUSTER_PREF}' or '{LABELS_PREF}foo' or '{ANNOTATIONS_PREF}foo' "
+                    f"or contain patters like: '${CLUSTER_PREF}'/'${LABELS_PREF}foo'/"
+                    f"'${ANNOTATIONS_PREF}foo'"
+                )
+                raise ValueError(err_msg)
+        return v
+
+    @classmethod
+    def normalize_key_string(cls, s: str) -> str:
+        return s.replace("/", "_").replace(".", "_").replace("-", "_")
+
+    @classmethod
+    def normalize_dict_keys(cls, metadata: Dict) -> Dict:
+        result = defaultdict(lambda: MISSING)
+        result.update({cls.normalize_key_string(k): v for k, v in metadata.items()})
+        return result
+
+    # if prefix not present, return ""
+    # else, if found, return replacement else return MISSING
+    @classmethod
+    def get_replacement(cls, prefix: str, value: str, normalized_replacements: Dict) -> str:
+        if prefix in value:  # value is in the format of "$prefix" or "prefix"
+            value = cls.normalize_key_string(value.replace(prefix, ""))
+            if "$" in value:
+                return Template(value).safe_substitute(normalized_replacements)
+            else:
+                return normalized_replacements[value]
+        return ""
+
+    @classmethod
+    def replace_token(cls, pattern: regex.Pattern, prefix: str, channel: str, replacements: Dict[str, str]) -> str:
+        tokens = pattern.findall(channel)
+        for token in tokens:
+            clean_token = token.replace("{", "").replace("}", "")
+            replacement = cls.get_replacement(prefix, clean_token, replacements)
+            if replacement:
+                channel = channel.replace(token, replacement)
+        return channel
+
+    @classmethod
+    def template(
+        cls,
+        channel_override: Optional[str],
+        default_channel: str,
+        cluster_name: str,
+        labels: Dict[str, str],
+        annotations: Dict[str, str],
+    ) -> str:
+        if not channel_override:
+            return default_channel
+
+        channel = channel_override
+        if CLUSTER_PREF in channel:
+            # replace "cluster_name" or "$cluster_name" or ${cluster_name} with the value of the cluster name
+            channel = CLUSTER_PREF_PATTERN.sub(cluster_name, channel)
+
```

Comment:
Since the below string manipulations might be expensive, performance wise, and are performed for each message, maybe it's worth leaving lines 85-92 under this `if`:
`if LABELS_PREF in channel or ANNOTATIONS_PREF in channel:`

We can skip it all, if the above is false

Reply:
72486a12dbadd1d36d52582d36ac547e5131843a